plugin-server/src/worker/ingestion/event-pipeline/cookielessServerHashStep.ts

```diff
@@ -0,0 +1,534 @@
+    timestamp: number,
+    eventTimeZone: string | undefined,
+    teamTimeZone: string
+): Date {
+    if (eventTimeZone) {
+        try {
+            return toStartOfDayInTimezone(timestamp, eventTimeZone)
+        } catch {
+            // pass
+        }
+    }
+    try {
+        return toStartOfDayInTimezone(timestamp, teamTimeZone)
+    } catch {
+        return toStartOfDayInTimezone(timestamp, TIMEZONE_FALLBACK)
+    }
+}
+
+export function createStatelessSessionId(
+    timestamp: number,
+    eventTimezone: string | undefined,
+    teamTimeZone: string,
+    hash: Uint32Array
+): UUID7 {
+    // A sessionId is a UUIDv7, which has a timestamp part and a random part. We need to find a deterministic way to
+    // generate this ID whilst meeting the requirements of posthog session IDs
+    // see https://posthog.com/docs/data/sessions#custom-session-ids
+
+    // For the timestamp part, use the start of the day, in this user's timezone
+    const timestampOfStartOfDay = toStartOfDayInTimezoneSafe(timestamp, eventTimezone, teamTimeZone).getTime()
+
+    // For the random part, use the first 10 bytes of the hash (74 bits are actually used), as a way of ensuring
+    // determinism with no state
+    const fakeRandomBytes = Buffer.from(hash.buffer).subarray(0, 10)
+
+    return new UUID7(timestampOfStartOfDay, fakeRandomBytes)
+}
+
+export function stripPIIProperties(event: PluginEvent) {
+    if (event.properties) {
+        // we use these properties in the hash, but they should not be written to disk if explicit consent was not given
+        delete event.properties['$ip']
+        delete event.properties['$raw_user_agent']
+        delete event.properties[COOKIELESS_EXTRA_HASH_CONTENTS_PROPERTY]
+    }
+    return event
+}
+
+interface SessionState {
+    sessionId: UUID7 // 16 bytes
+    // 8 bytes, LE uint64 (I couldn't bring myself to store a timestamp as a double, even in JS where everything is a double)
+    // log2(Date.now()) ~= 40, so 52 bits of mantissa *would* be fine, I'm just being an unreasonable pedant
+    lastActivityTimestamp: number
+}
+export function bufferToSessionState(buffer: Buffer): SessionState {
+    // the first 16 bytes are the session id
+    const sessionId = new UUID7(buffer.subarray(0, 16))
+    // the next 8 bytes are the last activity timestamp
+    const lastActivityTimestamp = Number(buffer.readBigUInt64LE(16))
+
+    return { sessionId, lastActivityTimestamp }
+}
+
+export function sessionStateToBuffer({ sessionId, lastActivityTimestamp }: SessionState): Buffer {
+    const buffer = Buffer.alloc(24)
+    buffer.set(sessionId.array, 0)
+    buffer.writeBigUInt64LE(BigInt(lastActivityTimestamp), 16)
+    return buffer
+}
+
+export function deleteExpiredSalts(): void {
+    for (const key in localSaltMap) {
+        if (!isCalendarDateValid(key)) {
+            delete localSaltMap[key]
+        }
+    }
+}
+
+setInterval(deleteExpiredSalts, DELETE_EXPIRED_SALTS_INTERVAL_MS)
```

Comment:
Maybe a comment would be nice here?

BTW, it's crazy that we just call this here instead of having a central place where we schedule everything

Reply:
It's now in a central place :)